### ISSUE TITLE
Reduce CPU usage

### ISSUE THREAD

**Issue by Jesse-V** [Jesse-V/Folding-Atomata#44]
Right now FoldingAtomata uses between 75% and 100% of a CPU core.
FAHViewer, by contrast, uses 15%.

The CPU usage may be coming from SlotViewer::update, which calculates the positions and orientations of all the bonds and atoms at every frame. A profiler may be needed to help see what's actually the heaviest bit.

Another possibility is to separate the update calls from the render calls by putting the update calls in another thread. That thread could run at a slower rate and/or a lower priority. Maybe there's a way to use some caching.

**Comment by Jesse-V** [comment]
After further consideration, I think making a separate slow update thread would be a good solution here. That also separates the updating from the rendering, and then the two can be controlled separately if need be.

**Comment by Jesse-V** [comment]
Actually, calling update() repeatedly from it's own thread seems to impact the CPU usage very little. Even when updating once every three seconds, 75% of the CPU is still used.

**Comment by Jesse-V** [comment]
I'm guessing that std::unordered_multimap in the rendering step may be the culprit here. Using vector< program > and vector< pair< programIndex, modelptr > > may be a better idea.

**Comment by Jesse-V** [comment]
I've done some profiling and the CPU usage stems from the repeated calls to Scene::render and all of it's subcalls. There are many, many calls to DataBuffer::enable/disable methods, but I need those. I've reduced the CPU usage as much as I can for now, but I can't escape the fact that I have to loop through thousands of models 60 times per second, and that requires work. Reducing the FPS does help though, but in my opinion the benefits of that are outweighed by the benefits of staying in sync with the monitor.

**Comment by Jesse-V** [comment]
I believe I can reduce CPU and simultaneously increase FPS by changing how I render all the objects. Something like https://www.opengl.org/wiki/Vertex_Rendering#Multi-Draw and https://www.opengl.org/wiki/Vertex_Rendering#Instancing may be the key here. Then I'll avoid calling glDrawElements and binding everything thousands of times per frame, which is likely the source of my problem.

**Comment by Jesse-V** [comment]
glDrawElementsInstanced is unavailable because it requires OpenGL 3.1. On my machine, Nouveau supplies OpenGL 3.0. My code uses 2.1, which does support glMultiDrawElements.

**Comment by Jesse-V** [comment]
http://stackoverflow.com/questions/7894739/opengl-drawing-the-same-polygon-many-times-in-multiple-places has some good ideas. Turns out glDrawElementsInstancedEXT is available under Nouveau drivers, at least on my machine.

Also, it might be better to put more load on the GPU. Instead of animated CPU-side and then taking the result of the GLM calculations and passing down the new atom positions, it might be a better idea to simply pass down the time as a uniform, and let the shaders do the interpolation in parallel. All they'd have to interpolate between is snapshot A and B, which could be sent down every three seconds, just like it's currently handled CPU-side.
